examples/dataflow-hashpipeline/hashpipeline.py

```diff
@@ -0,0 +1,213 @@
+       | 'Get DLP Findings' >> beam.ParDo(DlpFindingDoFn(opts.firestore_project, std_opts.runner))
+       | 'Determine if client SSN' >> beam.ParDo(ExistingSSNsDoFn(
+          opts.firestore_project, opts.salt, opts.secret_name, opts.collection_name, std_opts.runner))
+       | 'Apply windowing' >> beam.WindowInto(FixedWindows(5))
+       | 'Group by Filename' >> beam.GroupByKey()
+       | 'Count SSNs by Filename' >> beam.Map(lambda elem : (elem[0], sum(elem[1])))
+       | 'Format Message' >> beam.Map(fmt)
+       | 'Write to Pubsub' >> WriteToPubSub(opts.output_topic))
+
+class DlpFindingDoFn(beam.DoFn):
+  """Fetch DLP Findings as a PCollection"""
+  def __init__(self, project, runner):
+    beam.DoFn.__init__(self)
+    self.project = project
+    self.runner = runner
+    self.inspect_config = {
+      'info_types'    : [{'name': 'US_SOCIAL_SECURITY_NUMBER'}],
+      'min_likelihood': 'VERY_UNLIKELY',
+      'include_quote' : True # We need the output to match against the KV Store
+    }
+
+  def setup(self):
+    from google.cloud.dlp import DlpServiceClient
+    self.dlp_client = DlpServiceClient()
+
+  def process(self, element):
+    # TODO: Remove when version 2.22.0 is released BEAM-7885
+    if self.runner == 'DirectRunner':
+      self.setup()
+    # Convert the project id into a full resource id.
+    parent = self.dlp_client.project_path(self.project)
+    filename, chunk = element
+    # Call the API.
+    response = self.dlp_client.inspect_content(parent, self.inspect_config, {'value': chunk})
+
+    if response.result.findings:
+      for f in response.result.findings:
+        yield (filename, f.quote)
+
+class ExistingSSNsDoFn(beam.DoFn):
+  """Attempt to submit the finding to Firestore to see if the SSN exists"""
+  def __init__(self, project, salt, secret_name, collection_name, runner):
+    beam.DoFn.__init__(self)
+    self.project = project
+    self.salt = salt
+    self.secret_name = secret_name
+    self.collection_name = collection_name
+    self.runner = runner
+
+  def setup(self): 
+    import os
+    from google.cloud import firestore
+    from google.cloud import secretmanager
+    os.environ["GCLOUD_PROJECT"] = self.project
+    self.db = firestore.Client()
+    self.sm = secretmanager.SecretManagerServiceClient()
+
+  def process(self, element):
+    from base64 import b64decode
+    import hmac
+    # TODO: Remove when version 2.22.0 is released BEAM-7885
+    if self.runner == 'DirectRunner':
+      self.setup()
+    
+    filename, ssn = element
+    version = self.sm.access_secret_version('{}/versions/latest'.format(self.secret_name))
+    key = b64decode(version.payload.data)
+    
+    norm_ssn = ssn.replace('-', '')
+    salt = self.salt.encode('utf-8')
+    mac = hmac.new(key, msg=salt, digestmod='sha256')
+    mac.update(norm_ssn.encode('utf-8'))
+    digest = mac.hexdigest()
+ 
+    col = self.db.collection(self.collection_name)
+    doc = col.document(digest).get()
```

Comment:
This may pose a security risk since the SSN would have to be stored in plain text in the cache, which could be swapped to disk. A few attack vectors would be
1. If someone has compute.admin they could SSH into the host and see the SSN information in the cache by dumping memory
2. I'm not sure if the swap partition is saved when creating a machine image, but if it is, it could lead to the SSN being saved in an image
3. If cache is not cleared frequently, eventually a significant portion of the SSN dictionary could end up in the cache in plain text.

Reply:
> This may pose a security risk since the SSN would have to be stored in plain text in the cache, which could be swapped to disk. A few attack vectors would be

Disagree. I don't think we're opening any new attack vectors. 
You'd be implementing the caching logic yourself so this could be a design consideration.
Your cache could very easily cache the hashed values (which is what you look up anyway)

> 
> 1. If someone has compute.admin they could SSH into the host and see the SSN information in the cache by dumping memory

This should be mitigated by having port 22 closed in a firewall rule on the subnet where you run dataflow workers (for this job). It's very rare that any human / SA needs to ssh to these hosts and this should only be necessary during an incident (the process should be only open port 22 temporarily for dataflow workers during incident response w/ ticket).

If you cache hashed value they need compute.admin access and access to the key (which I suppose arguably is also in memory).

Trying to secure in-memory resources is pretty futile. If someone has compute admin, SSH access, and the ability to dump memory what's to stop them from doing this regularly to and intercepting the value of every successful lookup? At some point (e.g. L149) the document containing the hashed SSN is in memory for comparison and could be heap dumped.

It doesn't feel to me like we're increasing attack surface by caching.
The security consequences are in a breach of the VM a single heap dump may contain more hashed SSNs than if you do the lookup for each element.  

[Reference](https://g3doc.corp.google.com/company/gfw/support/cloud/playbooks/dataflow/out_of_memory_exceptions.md?cl=head#option-2-obtain-heap-dumps-from-a-running-not-yet-oomd-worker) (google internal) about getting heap dumps from workers.

> 2. I'm not sure if the swap partition is saved when creating a machine image, but if it is, it could lead to the SSN being saved in an image

This does not happen. 
Dataflow runs the [COS image](https://cloud.google.com/container-optimized-os/docs) and all user code is sandboxed i the sdk harness image. This has some [details you can review on the filesystem](https://cloud.google.com/container-optimized-os/docs/concepts/security#filesystem).

> 3. If cache is not cleared frequently, eventually a significant portion of the SSN dictionary could end up in the cache in plain text.

As the author of this code, you can control how frequently you clear the cache (based on # of elements in the cache there are easy to use tools in python e.g. [functools.lru_cache](https://docs.python.org/3/library/functools.html#functools.lru_cache) or [cachetools.TTLCache](https://pypi.org/project/cachetools/)).